### PR TITLE
Accept WebSocket protocols as string or sequence

### LINKED ISSUE #9053: Accept a sequence of strings in the WebSocket constructor
The WebIDL definitions for WebSocket has been updated upstream at the HTML specification documentation. In our current code, we use `optional (DOMString or DOMString) protocols` as the second argument to the constructor of WebSockets. The updated spech [here](https://html.spec.whatwg.org/multipage/comms.html#the-websocket-interface) instead uses `optional (DOMString and sequence<DOMString>) protocols`.

Code: `components/script/dom/webidls/WebSockets.webidl`, `components/script/dom/websocket.rs`
Spec: https://html.spec.whatwg.org/multipage/comms.html#the-websocket-interface
Test: `tests/wpt/web-platform-tests/websockets/constructor.html`

### PR DESCRIPTION
**Pull request by jsanders**
Fixes #9053

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/9360)

### CODE REVIEW

File: components/script/dom/websocket.rs [hunk 1]
```diff
@@ -193,9 +193,13 @@ impl WebSocket {
         }
 
         // Step 4.
-        let protocols: &[DOMString] = protocols
-                                      .as_ref()
-                                      .map_or(&[], |ref string| ref_slice(string));
+        let protocols: Vec<String> = match protocols {
```
Comment (KiChjang) [inline components/script/dom/websocket.rs]: I don't think the type annotation is necessary here.

File: components/script/dom/websocket.rs [hunk 2]
```diff
@@ -193,9 +193,13 @@ impl WebSocket {
...
+            Some(eString(string)) => vec![String::from(string)],
+            Some(eStringSequence(sequence)) => {
+                sequence.into_iter().map(|string| String::from(string)).collect()
```
Comment (KiChjang) [inline components/script/dom/websocket.rs]: Does `.map(String::from)` work here?

**Comment by KiChjang** [comment]
With my comments addressed, please squash and we'll merge this!

**Comment by jsanders** [comment]
@KiChjang: Done!

**Comment by KiChjang** [comment]
```
Ran 3837 tests finished in 227.0 seconds.
  • 3832 ran as expected. 691 tests skipped.
  • 5 tests crashed unexpectedly

Tests with unexpected results:
  ▶ CRASH [expected OK] /websockets/Create-Secure-valid-url-array-protocols.htm

  ▶ CRASH [expected OK] /websockets/Create-protocols-repeated.htm

  ▶ CRASH [expected OK] /websockets/Create-valid-url-array-protocols.htm

  ▶ CRASH [expected OK] /websockets/constructor/021.html

  ▶ CRASH [expected OK] /websockets/constructor/022.html
```

**Comment by ecoal95** [comment]
Hrrm, Got them all passing locally with this patch...

@KiChjang Do you know if there's a way to see the stack trace on the builders? I can't find any in the wpt log.

**Comment by jdm** [comment]
Unfortunately we don't have any automation to report backtraces of actual segfaults. These would need to be run under gdb either locally or on the build machine to get that information.

**Comment by ecoal95** [comment]
Just for the record, while I dig into this, the segfault is on spidermonkey, which is a bit worrying, and just on release builds (which is even more).

At least it seems deterministic... There it goes my backtrace:

```
0x0000555556738f1e in js::GetLengthProperty (cx=0x7fff9c027340, obj=..., lengthp=0x7fffc7fe702c)                                                                                              
    at /home/emilio/projects/moz/servo/.cargo/git/checkouts/mozjs-06d7f04b6dbb8a8e/master/mozjs/js/src/jsarray.cpp:63
63      if (obj->is<ArrayObject>()) {
(gdb) bt
#0  0x0000555556738f1e in js::GetLengthProperty (cx=0x7fff9c027340, obj=..., lengthp=0x7fffc7fe702c)
    at /home/emilio/projects/moz/servo/.cargo/git/checkouts/mozjs-06d7f04b6dbb8a8e/master/mozjs/js/src/jsarray.cpp:63
#1  0x0000555555efb952 in conversions::Vec$LT$T$GT$.FromJSValConvertible::from_jsval::h11377707114978210586 ()
#2  0x0000555555efaf01 in dom::bindings::codegen::UnionTypes::StringOrStringSequence.FromJSValConvertible::from_jsval::heefa2eec5fa5708eJhx ()
#3  0x0000555555f53c0b in dom::bindings::codegen::Bindings::WebSocketBinding::_constructor::hb419dab516bee77dxSi ()
```

**Comment by jdm** [comment]
Make sure you're building with debug mozjs enabled in your .servobuild to obtain maximal helpful assertions.

**Comment by ecoal95** [comment]
Heh, you'll like this, @jdm:

in [rust-mozjs/src/conversions.rs](https://github.com/servo/rust-mozjs/blob/master/src/conversions.rs#L506):

**Segfault**:

```rust
let handle_obj = Handle::from_marked_location(&value.to_object());
```

**No segfault** (nor triggered assertions):

```rust
let object = value.to_object();
assert!(!object.is_null());
let handle_obj = Handle::from_marked_location(&object);
```

So... Maybe optimisations are playing a big role here (I guess the assert prevents some kind of optimisation).

You can try this (in `rust-mozjs`) with: `cargo test --release test_vec_conversion`.

**Comment by ecoal95** [comment]
BTW: A servo build with debug-mozjs enabled gives me this on startup, independently of the page loaded of course.

```
$ ./target/release/servo     
Assertion failure: !mEntered, at ../../dist/include/mozilla/ReentrancyGuard.h:39
```

**Comment by jdm** [comment]
There is a reason I give every use of Handle::from_marked_location extreme side-eye. As established over IRC, the handle is outliving the stack `JSObject*` that it contains a pointer to, so the JSAPI code receives garbage.

**Comment by jsanders** [comment]
@ecoal95 / @jdm: D'oh, I guess I probably only tested a debug build locally. Seems like you guys have a better handle on this than me, but let me know if there's anything I need to do to fix it.

**Comment by ecoal95** [comment]
@jsanders Actually it was my fault. This will have to wait until a `rust-mozjs` update lands, but the code is fine :P

**Comment by jsanders** [comment]
@ecoal95: Ok no problem!

**Comment by KiChjang** [comment]
rust-mozjs update has landed, let's land this.

@bors-servo retry

**Pull request merged**